### PR TITLE
feat: add slot to support extra position

### PR DESCRIPTION
**Pull request by jesse3mh9a**
https://github.com/ant-design/ant-design/issues/24522#issue-626162474

### Tabs

| name | type | default | description |
| --- | --- | --- | --- |
| tabBarExtraContent | `React.ReactNode` \| `{left: React.ReactNode, right: React.ReactNode}` |  `-` |

### CODE REVIEW

**Comment by afc163** [comment]
不如直接支持 `tabBarExtraContent={{ left: <div />, right: <div /> }}`.

**Comment by jesse3mh9a** [comment]
@afc163   👌 已经修改了一下

File: src/TabNavList/index.tsx [hunk 1]
```diff
@@ -44,6 +47,40 @@ export interface TabNavListProps {
   children?: (node: React.ReactElement) => React.ReactElement;
 }
 
+
+interface ExtraContentProps {
+  name: TabBarExtraPosition;
```
Comment (07akioni) [inline src/TabNavList/index.tsx:52]: name it as `position` or `placement`, not `name`

File: src/TabNavList/index.tsx [hunk 2]
```diff
@@ -44,6 +47,40 @@ export interface TabNavListProps {
...
+  }
+
+  const components: TabBarExtraMap = {
+    left: null,
+    right: null,
+  };
+
+  const assertExtra = extra as TabBarExtraMap;
+
+  const { left = null, right = null } = assertExtra || {};
```
Comment (07akioni) [inline src/TabNavList/index.tsx:69]: no need for `|| {}`, since guard is set at the start of the function

File: src/TabNavList/index.tsx [hunk 3]
```diff
@@ -44,6 +47,40 @@ export interface TabNavListProps {
...
+
+  if(left === null && right === null) {
+    // default right
+    components.right = extra;
+  } else {
+    components.left = left;
+    components.right = right;
+  }
+
+  const com = components[name];
```
Comment (07akioni) [inline src/TabNavList/index.tsx:79]: `content` is better than `com`?

File: src/TabNavList/index.tsx [hunk 4]
```diff
@@ -437,7 +476,7 @@ function TabNavList(props: TabNavListProps, ref: React.Ref<HTMLDivElement>) {
         className={!hasDropdown && operationsHiddenClassName}
       />
 
-      {extra && <div className={`${prefixCls}-extra-content`}>{extra}</div>}
+    <ExtraContent name="right" extra={extra} prefixCls={prefixCls}/>
```
Comment (07akioni) [inline src/TabNavList/index.tsx:479]: indent?

Comment (07akioni) [reply src/TabNavList/index.tsx:479]: `prefixCls={prefixCls}/>` -> `prefixCls={prefixCls} />`

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx]: Done!
Thanks for your advice

File: src/TabNavList/index.tsx [hunk 5]
```diff
@@ -437,7 +476,7 @@ function TabNavList(props: TabNavListProps, ref: React.Ref<HTMLDivElement>) {
         className={!hasDropdown && operationsHiddenClassName}
       />
 
-      {extra && <div className={`${prefixCls}-extra-content`}>{extra}</div>}
+    <ExtraContent position="right" extra={extra} prefixCls={prefixCls} />
```
Comment (07akioni) [inline src/TabNavList/index.tsx:479]: indent here is still wired

Comment (07akioni) [reply src/TabNavList/index.tsx:479]: it should be 6 spaces

Comment (afc163) [reply src/TabNavList/index.tsx:479]:
```suggestion
      <ExtraContent position="right" extra={extra} prefixCls={prefixCls} />
```

**Comment by jesse3mh9a** [comment]
@afc163  @07akioni 
update example

当`tabBarExtraContent` 为 `{}`会报错, 
 如果改成
```
try {
      React.Children.count(extra)
      components.right = extra;
    } catch (error) {
      console.log(error);
    }
```

会有`Objects are not valid as a React child...`相关提示,

**Comment by 07akioni** [comment]
CI Failed?

File: src/TabNavList/index.tsx [hunk 6]
```diff
@@ -44,6 +47,46 @@ export interface TabNavListProps {
...
+  };
+
+  const assertExtra = extra as TabBarExtraMap;
+
+  const { left = null, right = null } = assertExtra;
+
+  if (left === null && right === null) {
+    // default right
+    try {
+      React.Children.count(extra)
```
Comment (07akioni) [inline src/TabNavList/index.tsx:74]: why count extra?

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx:74]: if `extra` is `{}`, will throw error
I'm not so sure here, can you give me some advice

**Comment by jesse3mh9a** [comment]
CI passed ✅ @07akioni  but `codecov/patch`?

File: src/TabNavList/index.tsx (hunk 6 above)
Comment (07akioni) [reply src/TabNavList/index.tsx]: 为什么需要 count extra 呢？这里我其实没搞明白，这个语句是干什么的

**Comment by 07akioni** [comment]
> CI passed ✅ @07akioni but `codecov/patch`?

see the link. some lines changed are not covered in tests

https://codecov.io/gh/react-component/tabs/pull/286/diff?src=pr&el=tree#diff-c3JjL1RhYk5hdkxpc3QvaW5kZXgudHN4

**Comment by jesse3mh9a** [comment]
@07akioni  @afc163    `codecov/patch`  这个 我不是很清楚 能帮我看下 什么问题吗

**Comment by 07akioni** [comment]
> @07akioni @afc163 `codecov/patch`  这个 我不是很清楚 能帮我看下 什么问题吗

https://codecov.io/gh/react-component/tabs/compare/36e0a77dd5aa26b0b70b237453fb962585168aa5...23d694476d7e0eb9040bd8f6e4f4a59c9a9762c1/diff

第 75，76 行没有覆盖

**Comment by jesse3mh9a** [comment]
@07akioni @afc163  update test, all passed✅

File: src/TabNavList/index.tsx [hunk 7]
```diff
@@ -45,6 +48,55 @@ export interface TabNavListProps {
...
+
+const isNullish = (value: any) => value === undefined || value === null
+
+const isReactText = (value: any) => typeof value === 'string' || typeof value === "number"
+
+const isReactElement = (value: any) => React.isValidElement(value)
+
+const isReactChild = (value: any) => isReactText(value) || isReactElement(value)
+
+const ExtraContent = ({ position, prefixCls, extra }: ExtraContentProps) => {
```
Comment (07akioni) [inline src/TabNavList/index.tsx:68]: 这个函数用处理这么复杂吗？感觉原先逻辑就差不多呀，这里类型不用 runtime 检验的，声明好类型不需要检验的

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx:68]: `examples/extra`中动态分配left rigth,如果都为空, 很多业务场景里面 会得到一个`{}`,这样会产生一个错误.
所以这里我想更直接的处理预期定义的内容

Comment (07akioni) [reply src/TabNavList/index.tsx:68]: 得到一个 {} 就不显示呗？逻辑上说得通

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx:60]: 嗯 我想了下 是可以简单点 我revert了, 更新了下example

File: src/TabNavList/index.tsx [hunk 8]
```diff
@@ -45,6 +48,47 @@ export interface TabNavListProps {
...
+  if (extra === undefined) return null;
+
+  const components: TabBarExtraMap = {};
+
+  const assertExtra = extra as TabBarExtraMap;
+
+  const extraHasKey = (key: TabBarExtraPosition): boolean => {
+    // Cannot convert undefined or null to object (from travis-ci)
+    if (assertExtra === null || assertExtra === undefined) return false
```
Comment (07akioni) [inline src/TabNavList/index.tsx:69]: 这里应该在上面 guard，写在 line 61

File: src/TabNavList/index.tsx [hunk 9]
```diff
@@ -45,6 +48,47 @@ export interface TabNavListProps {
...
+
+    return hasOwn.call(assertExtra, key)
+  }
+
```
Comment (07akioni) [inline src/TabNavList/index.tsx:73]:
```typescript
if (!extra) return null

if (position === 'right') {
  return extra.right || extra || null
}

if (position === 'left') {
  return extra.left || null
}
```

Comment (07akioni) [reply src/TabNavList/index.tsx:73]: 我感觉想表达的是这个逻辑？

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx:73]: - `{ left: <div></div> }` 
- `{ right: <div></div> }` 
- `{ left:<div></div>, right:<div></div> }` 
- `<div></div>`

 api可以接收上面任意一个,

```
{ left: <div></div> }

if (position === 'right') {
  return extra.right || extra || null
}
```
 这个会 返回 `extra={left: <div></div>}` 而不是`ReactNode`

Comment (07akioni) [reply src/TabNavList/index.tsx:73]:
```
if (!extra) return null

if (!extra.left && !extra.right) return extra

if (position === 'right') return extra.right || null

if (position === 'left') return extra.left || null
```

Comment (jesse3mh9a) [reply src/TabNavList/index.tsx]: 👌Done

**Comment by afc163** [comment]
https://github.com/react-component/tabs/releases/tag/v11.6.0

**Pull request merged**